rust/syntect-plugin/src/main.rs

```diff
@@ -210,82 +219,175 @@ impl<'a> Syntect<'a> {
+                "insert" => {
+                    let range = region.new_offset..region.new_offset + region.len;
+                    let is_whitespace = {
+                        let insert_region =
+                            view.get_region(range).expect("view must return region");
+                        insert_region.as_bytes().iter().all(u8::is_ascii_whitespace)
+                    };
+                    if !is_whitespace {
+                        self.check_indent_active_edit(view, line_of_edit)
+                    } else {
+                        Ok(())
+                    }
                 }
-
-                let indent = self.indent_for_next_line(line, use_spaces, tab_size);
-                let ix = start + text.len();
-                let interval = Interval::new(ix, ix);
-                //TODO: view should have a `get_edit_builder` fn?
-                let mut builder = DeltaBuilder::new(buf_size);
-                builder.replace(interval, indent.into());
-
-                let delta = builder.build();
-                Some(delta)
-            } else {
-                None
+                other => panic!("unexpected edit_type {}", other),
             };
 
-            if let Some(delta) = result {
-                view.edit(delta, INDENTATION_PRIORITY, false, false, String::from("syntect"));
+            if let Err(e) = result {
+                eprintln!("error in autoindent {:?}", e);
             }
         }
     }
 
-    /// Returns the string which should be inserted after the newline
-    /// to achieve the desired indentation level.
-    fn indent_for_next_line<'b>(
-        &self,
-        prev_line: &'b str,
-        use_spaces: bool,
-        tab_size: usize,
-    ) -> Cow<'b, str> {
-        let leading_ws = prev_line
-            .char_indices()
-            .find(|&(_, c)| !c.is_whitespace())
-            .or_else(|| prev_line.char_indices().last())
-            .map(|(idx, _)| unsafe { prev_line.get_unchecked(0..idx) })
-            .unwrap_or("");
-
-        if self.increase_indentation(prev_line) {
-            let indent_text =
-                if use_spaces { &"                                    "[..tab_size] } else { "\t" };
-            format!("{}{}", leading_ws, indent_text).into()
+    /// Called when freshly computing a line's indent level, such as after
+    /// a newline, or when reindenting a block.
+    fn autoindent_line(&mut self, view: &mut MyView, line: usize) -> Result<(), Error> {
+        let _t = trace_block("Syntect::autoindent", &["syntect"]);
+        debug_assert!(line > 0);
+        let tab_size = view.get_config().tab_size;
+        let current_indent = self.indent_level_of_line(view, line);
+        let base_indent = self
+            .previous_nonblank_line(view, line)?
+            .map(|l| self.indent_level_of_line(view, l))
+            .unwrap_or(0);
+        let increase_level = self.test_increase(view, line)?;
```

Comment:
If I test here whether it should be decreased with `test_decrease` it returns an error (`error in autoindent PeerDisconnect`). The reason for this seems to be that at some point in `get_metadata` were are trying to get the `line` with `view.get(line)`. View seems to access some cache with the lines but cannot find `line` and returns an error. I am not sure if that is the expected behaviour?